File: api/BeSureApi/Models/Invoice.cs
namespace BeSureApi.Models
{
    public class InvoiceScheduleList
    {
            public int Id { get; set; }
            public int ContractId { get; set; }
            public string ContractNumber { get; set; }
            public string CustomerName { get; set; }
            public string InvoiceNumber { get; set; }
            public DateTime StartDate { get; set; }
            public DateTime EndDate { get; set; }
            public DateTime ScheduledInvoiceDate { get; set; }
            public decimal ScheduledInvoiceAmount { get; set; }
            public decimal? NetInvoiceAmount { get; set; }
            public decimal? CollectedAmount { get; set; }
            public int ContractInvoiceId { get; set; }
            public bool IsInvoiceApproved { get; set; }
    }

    public class InvoiceCollectionDetail
    {
        public int Id { get; set; }
        public DateTime InvoiceDate { get; set; }
        public decimal NetInvoiceAmount { get; set; }
        public decimal CollectedAmount { get; set; }
        public decimal OutstandingAmount { get; set; }
        public decimal TdsDeductedAmount { get; set; }
        public decimal TdsPaidAmount { get; set; }
        public decimal GstTdsDeductedAmount { get; set; }
        public decimal GstTdsPaidAmount { get; set; }
        public decimal PenaltyAmount { get; set; }
        public decimal SecurityDepositAmount { get; set; }
        public decimal CustomerExpenseAmount { get; set; }
        public decimal OtherDeductionAmount { get; set; }
        public string? OtherDeductionRemarks { get; set; }
        public decimal WriteOffAmount { get; set; }
        public string? WriteOffAmountRemarks { get; set; }
        public int PrimaryAccountableId { get; set; }
        public int SecondaryAccountableId { get; set; }
        public string InvoiceNumber { get; set; }
        public int CreatedBy { get; set; }
        public DateTime CreatedOn { get; set; }
        public int? ModifiedBy { get; set; }
        public DateTime? ModifiedOn { get; set; }
        public string CustomerName { get; set; }
        public string ContractNumber { get; set; }
    }

    public class ReceiptListForInvoice
    {
        public decimal MappedAmount { get; set; }
        public string ReceiptNumber { get; set; }
        public decimal InvoiceAmount { get; set; }
        public decimal ReceiptAmount { get; set; }
        public DateTime ReceiptDate { get; set; }
    }
    public class InvoiceCollectionViewWithReceipt
    {
        public InvoiceCollectionDetail InvoiceDetail { get; set; }
        public List<ReceiptListForInvoice>? InvoiceReceiptList { get; set; }
    }
}

File: ui/src/types/invoice.ts
import { Decoder, array, boolean, nullable, number, object, string } from "decoders";

export interface InvoiceScheduleListDetails {
    Id: number;
    InvoiceNumber: string|null;
    CustomerName:string;
    ContractNumber:string;
    StartDate: string;
    EndDate: string;
    ScheduledInvoiceDate: string;
    ScheduledInvoiceAmount: number;
    NetInvoiceAmount:number|null;
    CollectedAmount:number|null;
    ContractInvoiceId: number;
    IsInvoiceApproved: boolean;
  }
  
  export const invoiceScheduleListDetailsDecoder: Decoder<InvoiceScheduleListDetails> = object({
    Id: number,
    InvoiceNumber: nullable(string),
    CustomerName:string,
    ContractNumber:string,
    StartDate: string,
    EndDate: string,
    ScheduledInvoiceDate: string,
    ScheduledInvoiceAmount: number,
    NetInvoiceAmount:nullable(number), 
    CollectedAmount:nullable(number),
    ContractInvoiceId: number,
    IsInvoiceApproved: boolean
  });
  
  export interface InvoiceScheduleList {
    InvoiceScheduleList: InvoiceScheduleListDetails[];
    TotalRows: number;
    PerPage: number;

  }
  
  export const invoiceScheduleListDecoder: Decoder<InvoiceScheduleList> = object({
    InvoiceScheduleList: array(invoiceScheduleListDetailsDecoder),
    TotalRows: number,
    PerPage: number
  });

  export interface InvoiceSearchForFilter {
	SearchText: string;
	StartDate: string;
	EndDate: string;
}

export interface InvoiceCollectionDetail {
  Id: number|null;
  NetInvoiceAmount: number|null;
  CollectedAmount:number|null;
  OutstandingAmount:number|null;
  TdsDeductedAmount:number|null;
  TdsPaidAmount:number|null;
  GstTdsDeductedAmount:number|null;
  GstTdsPaidAmount:number|null;
  OtherDeductionAmount:number|null;
  CustomerExpenseAmount:number|null;
  SecurityDepositAmount:number|null;
  PenaltyAmount:number|null;
  WriteOffAmount:number|null;
  InvoiceNumber:string|null;
  CustomerName:string|null; 
  ContractNumber:string|null;
  InvoiceDate:string|null
  }
  
  export const invoiceCollectionDetailDecoder: Decoder<InvoiceCollectionDetail> = object({
  Id: number,
  NetInvoiceAmount:number,
  CollectedAmount:number,
  OutstandingAmount:number,
  TdsDeductedAmount:number,
  TdsPaidAmount:number,
  GstTdsDeductedAmount:number,
  GstTdsPaidAmount:number,
  OtherDeductionAmount:number,
  CustomerExpenseAmount:number,
  SecurityDepositAmount:number,
  PenaltyAmount:number,
  WriteOffAmount:number,
  InvoiceNumber:string,
  CustomerName:string,
  ContractNumber:string,
  InvoiceDate:nullable(string)
  });

export interface ReceiptDetailForInvoiceCollection {
  MappedAmount: number;
  ReceiptAmount: number;
  ReceiptNumber: string;
  InvoiceAmount: number;
  ReceiptDate : string;
}

export const receiptDetailForInvoiceCollectionDecoder: Decoder<ReceiptDetailForInvoiceCollection> = object({
  MappedAmount: number,
  ReceiptAmount: number,
  ReceiptNumber: string,
  InvoiceAmount: number,
  ReceiptDate :string
});

export interface InvoiceCollectionView {
  InvoiceDetail: InvoiceCollectionDetail;
  InvoiceReceiptList: ReceiptDetailForInvoiceCollection[]
}

export const invoiceCollectionViewDecoder: Decoder<InvoiceCollectionView> = object({
  InvoiceDetail: invoiceCollectionDetailDecoder,
  InvoiceReceiptList: array(receiptDetailForInvoiceCollectionDecoder)
});